x/precisebank/keeper/send.go

```diff
@@ -19,23 +32,291 @@ func (k Keeper) SendCoins(
+	// Full integer amount transfer, including direct transfer of borrow/carry
+	// if any.
+	if integerAmt.IsPositive() {
+		transferCoin := sdk.NewCoin(types.IntegerCoinDenom, integerAmt)
+		if err := k.bk.SendCoins(ctx, from, to, sdk.NewCoins(transferCoin)); err != nil {
+			return k.wrapError(ctx, from, amt, err)
+		}
+	}
+
+	// Case #2: Sender borrow, NO recipient carry
+	// Sender borrows by transferring 1 integer amount to reserve to account for
+	// lack of fractional balance.
+	if senderNeedsBorrow && !recipientNeedsCarry {
+		borrowCoin := sdk.NewCoin(types.IntegerCoinDenom, sdk.NewInt(1))
+		if err := k.bk.SendCoinsFromAccountToModule(
+			ctx,
+			from, // sender borrowing
+			types.ModuleName,
+			sdk.NewCoins(borrowCoin),
+		); err != nil {
+			return k.wrapError(ctx, from, amt, err)
+		}
+	}
+
+	// Case #3: NO sender borrow, recipient carry.
+	// Recipient's fractional balance carries over to integer balance by 1.
+	// Always send carry from reserve before receiving borrow from sender to
+	// ensure reserve always has sufficient balance starting from 0.
+	if !senderNeedsBorrow && recipientNeedsCarry {
+		carryCoin := sdk.NewCoin(types.IntegerCoinDenom, sdk.NewInt(1))
+		if err := k.bk.SendCoinsFromModuleToAccount(
+			ctx,
+			types.ModuleName,
+			to, // recipient carrying
+			sdk.NewCoins(carryCoin),
+		); err != nil {
+			// Panic instead of returning error, as this will only error
+			// with invalid state or logic. Reserve should always have
+			// sufficient balance to carry fractional coins.
+			panic(fmt.Errorf("failed to carry fractional coins to %s: %w", to, err))
+		}
+	}
+
+	// Update fractional balances
+	k.SetFractionalBalance(ctx, from, senderNewFracBal)
+	k.SetFractionalBalance(ctx, to, recipientNewFracBal)
+
+	return nil
+}
+
+// subFromFractionalBalance subtracts a fractional amount from the provided
+// current fractional balance, returning the new fractional balance and true if
+// an integer borrow is required.
+func subFromFractionalBalance(
+	currentFractionalBalance sdkmath.Int,
+	amountToSub sdkmath.Int,
+) (sdkmath.Int, bool) {
+	// Enforce that currentFractionalBalance is not a full balance.
+	if currentFractionalBalance.GTE(types.ConversionFactor()) {
+		panic("currentFractionalBalance must be less than ConversionFactor")
+	}
+
+	if amountToSub.GTE(types.ConversionFactor()) {
+		panic("amountToSub must be less than ConversionFactor")
+	}
+
+	newFractionalBalance := currentFractionalBalance.Sub(amountToSub)
+
+	// Insufficient fractional balance, so we need to borrow.
+	borrowRequired := newFractionalBalance.IsNegative()
+
+	if borrowRequired {
+		// Borrowing 1 integer equivalent amount of fractional coins. We need to
+		// add 1 integer equivalent amount to the fractional balance otherwise
+		// the new fractional balance will be negative.
+		newFractionalBalance = newFractionalBalance.Add(types.ConversionFactor())
+	}
+
+	return newFractionalBalance, borrowRequired
```

Comment:
Bit different from suggested @nddeluca though I think its clearer this way, by comparing with negative / conversionFactor amounts instead of previous